### PR TITLE
Add benchmarks

### PR DESCRIPTION
**Pull request by gautierronan**
Benchmarks for comparing dynamiqs vs qutip.

The structure is pretty straightforward: 
- a `models/` folder containing 
  - a base model class (`bench_model.py`)
  - several models to benchmark (`cat_cnot.py`, `driven_cavity.py`, ...)
- a notebook that runs all benchmarks and produces figures (`run_benchmarks.ipynb`)

Most of the logic is included in the `BenchModel` class. In particular the automatic runtime benchmarks. All is needed for users later on is:
1. to create their custom models by defining `init_qutip(...)` and `init_dynamiqs(...)` methods that initialize an arbitrary function to benchmark, and its input arguments as kwargs. Note that this works for any function, not just `sesolve`/`mesolve`, so we can also write parameter sweeps.
2. to initialize their model and call `model.benchmark(library, backend)` to get the average runtime of the corresponding function. 

Resulting plot when running on our RTX 4090. 

image
image

Related to DYN-53

### CODE REVIEW

**Comment by gautierronan** [comment]
Sorry for the set of messy commits: I was going back and forth between several implementations, but I think I'm happy with the structure now.

You can review file by file, it should be pretty straightforward. Most of the logic is now in the `bench_model.py` file.

File: benchmarks/models/catcnot.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    ):
+        # time evolution
+        alpha = jnp.sqrt(nbar)
+        gate_time = jnp.pi / (4 * alpha * g_cnot)
+        tsave = jnp.linspace(0.0, gate_time, num_tsave)
+
+        # operators
+        ac = dq.tensor(dq.destroy(N), dq.eye(N))
+        nt = dq.tensor(dq.eye(N), dq.number(N))
+        i = dq.tensor(dq.eye(N), dq.eye(N))
```
Comment (abocquet) [inline benchmarks/models/catcnot.py:66]: Nit: `i` is usually devoted to loop indexes, perhaps `I` or `eye` or `identity` would be more straightforward

**Review by pierreguilmin — COMMENTED** [review]
Coooool 😎 I'll try to find the time for a complete review soon. Few high-level comments:
- Can you edit the PR description with the recent changes? Why did you decide to move to a base class in the end? (I agree it's practical, but it also makes the benchmark code not very accessible for an outsider)
- To respect Python conventions, file names should be `catcnot.py` -> `cat_cnot.py`, `crgate.py` -> `cross_resonance_gate.py`, etc...

**Review by abocquet — CHANGES_REQUESTED** [review]
Running 2nd cell:
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 10
      7 print(f'Running benchmark for {model.__class__.__name__}')  # noqa: T201
      9 # check qutip and dynamiqs models are equal
---> 10 model.check_args()
     12 # benchmark models
     13 timer_qutip_cpu = model.benchmark('qutip', 'cpu')

File ~/Programmation/dynamiqs/benchmarks/models/bench_model.py:61, in BenchModel.check_args(self)
     59 """Check that all input arguments are equal between qutip and dynamiqs."""
     60 if 'H' in self.kwargs_qutip and 'H' in self.kwargs_dynamiqs:
---> 61     self._check_operators_equal(
     62         self.kwargs_qutip['H'], self.kwargs_dynamiqs['H']
     63     )
     64 if 'psi0' in self.kwargs_qutip and 'psi0' in self.kwargs_dynamiqs:
     65     self._check_operators_equal(
     66         self.kwargs_qutip['psi0'], self.kwargs_dynamiqs['psi0']
     67     )

File ~/Programmation/dynamiqs/benchmarks/models/bench_model.py:90, in BenchModel._check_operators_equal(self, op_qutip, op_dynamiqs)
     87 elif isinstance(op_qutip, list) and isinstance(op_dynamiqs, dq.TimeArray):
     88     # check that operators are equal when evaluated at t=0.0 and t=0.23113
     89     op_qutip = qt.QobjEvo(op_qutip)
---> 90     assert jnp.allclose(op_qutip(0.0).full(), op_dynamiqs(0.0))
     91     assert jnp.allclose(op_qutip(0.23113).full(), op_dynamiqs(0.23113))

File ~/.pyenv/versions/3.11.3/envs/dynamiqs/lib/python3.11/site-packages/qutip/qobjevo.py:641, in QobjEvo.__call__(self, t, data, state, args)
    639     op_t = self.cte.copy()
    640     for part in self.ops:
--> 641         op_t += part.qobj * part.get_coeff(t, self.args)
    643 return op_t

TypeError: CrossResonanceGate.init_qutip.<locals>.<lambda>() takes 1 positional argument but 2 were given
```

**Comment by gautierronan** [comment]
> Running 2nd cell:
> ```python
> ---------------------------------------------------------------------------
> TypeError                                 Traceback (most recent call last)
> Cell In[2], line 10
>       7 print(f'Running benchmark for {model.__class__.__name__}')  # noqa: T201
>       9 # check qutip and dynamiqs models are equal
> ---> 10 model.check_args()
>      12 # benchmark models
>      13 timer_qutip_cpu = model.benchmark('qutip', 'cpu')
> 
> File ~/Programmation/dynamiqs/benchmarks/models/bench_model.py:61, in BenchModel.check_args(self)
>      59 """Check that all input arguments are equal between qutip and dynamiqs."""
>      60 if 'H' in self.kwargs_qutip and 'H' in self.kwargs_dynamiqs:
> ---> 61     self._check_operators_equal(
>      62         self.kwargs_qutip['H'], self.kwargs_dynamiqs['H']
>      63     )
>      64 if 'psi0' in self.kwargs_qutip and 'psi0' in self.kwargs_dynamiqs:
>      65     self._check_operators_equal(
>      66         self.kwargs_qutip['psi0'], self.kwargs_dynamiqs['psi0']
>      67     )
> 
> File ~/Programmation/dynamiqs/benchmarks/models/bench_model.py:90, in BenchModel._check_operators_equal(self, op_qutip, op_dynamiqs)
>      87 elif isinstance(op_qutip, list) and isinstance(op_dynamiqs, dq.TimeArray):
>      88     # check that operators are equal when evaluated at t=0.0 and t=0.23113
>      89     op_qutip = qt.QobjEvo(op_qutip)
> ---> 90     assert jnp.allclose(op_qutip(0.0).full(), op_dynamiqs(0.0))
>      91     assert jnp.allclose(op_qutip(0.23113).full(), op_dynamiqs(0.23113))
> 
> File ~/.pyenv/versions/3.11.3/envs/dynamiqs/lib/python3.11/site-packages/qutip/qobjevo.py:641, in QobjEvo.__call__(self, t, data, state, args)
>     639     op_t = self.cte.copy()
>     640     for part in self.ops:
> --> 641         op_t += part.qobj * part.get_coeff(t, self.args)
>     643 return op_t
> 
> TypeError: CrossResonanceGate.init_qutip.<locals>.<lambda>() takes 1 positional argument but 2 were given
> ```
> 
> (done on a mac M1, no GPU available for JAX)

That's because of incompatibility between QuTiP 4.7 and QuTiP 5 🥲 Should we write these benchmarks for one or the other?

**Comment by abocquet** [comment]
I have no opinion but then maybe but a version check somewhere

**Comment by abocquet** [comment]
For information, [pypi-multi-versions](https://pypi.org/project/pypi-multi-versions/) can help handling two versions of a same package. Maybe this could be a solution.

**Comment by gautierronan** [comment]
Closing this. Will replace by an alternative version of the PR soon.

**Pull request closed without merging**